### PR TITLE
feat: mv max_flowcells on disk to settings file

### PR DESCRIPTION
**Pull request by ingkebil**
This PR moves the max_flowcells on disk setting to the settings file. If the setting in the settings file doesn't exist, it'l default to 1000 on disk. It also adds a dry-run cli option so we can actually test this.

**How to test**:
1. install on hasta stage
1. activate stage: `us`
1. run following command: `cg backup fetch-flowcell`

**Expected outcome**:
message 'maximum flowcells ondisk reached should NOT appear
Flowcell should be fetched from PDC: `<flowcell>: retreiving (sic) from PDC`

**Review:**
- [x] code approved by @patrikgrenfeldt 
- [x] tests executed by @barrystokman 
- [x] "Merge and deploy" approved by @moonso 
Thanks for filling in who performed the code review and the test!

This is |minor| **version bump** because it adds a feature to the system in a backwards compatible manner

### CODE REVIEW

**Comment by patrikgrenfeldt** [comment]
@ingkebil you seem to have written: "THIS PR IS STILL MISSING UNIT TESTS AND SHOULD NOT BE REVIEWED YET." 

What is the status?

**Comment by patrikgrenfeldt** [comment]
"THIS PR IS STILL MISSING UNIT TESTS AND SHOULD NOT BE REVIEWED YET.", Is this still the case?

**Comment by patrikgrenfeldt** [comment]
@ingkebil What should the fate of this PR be?

File: cg/meta/backup.py [hunk 1]
```diff
@@ -12,17 +12,18 @@
...
         self.pdc = pdc_api
+        self.max_flowcells = max_flowcells
 
-    def maximum_flowcells_ondisk(self, max_flowcells: int = 1000) -> bool:
+    def maximum_flowcells_ondisk(self) -> bool:
         """Check if there's too many flowcells already "ondisk"."""
         ondisk_flowcells = self.status.flowcells(status='ondisk').count()
-        LOG.debug("ondisk flowcells: %s", ondisk_flowcells)
-        return ondisk_flowcells > max_flowcells
+        LOG.debug(f"ondisk flowcells: {ondisk_flowcells}")
```
Comment (ingkebil) [inline cg/meta/backup.py]:
```suggestion
        LOG.debug("ondisk flowcells: %s", ondisk_flowcells")
```

File: cg/meta/backup.py [hunk 2]
```diff
@@ -12,17 +12,18 @@
...
         self.pdc = pdc_api
+        self.max_flowcells = max_flowcells
 
-    def maximum_flowcells_ondisk(self, max_flowcells: int = 1000) -> bool:
+    def maximum_flowcells_ondisk(self) -> bool:
         """Check if there's too many flowcells already "ondisk"."""
         ondisk_flowcells = self.status.flowcells(status='ondisk').count()
-        LOG.debug("ondisk flowcells: %s", ondisk_flowcells)
-        return ondisk_flowcells > max_flowcells
+        LOG.debug("ondisk flowcells: %s", ondisk_flowcells")
```
Comment (ingkebil) [inline cg/meta/backup.py]:
```suggestion
        LOG.debug("ondisk flowcells: %s", ondisk_flowcells)
```

**Comment by ingkebil** [comment]
up to you man

**Comment by barrystokman** [comment]
Before:
image

After:
image
:heavy_check_mark:

**Comment by barrystokman** [comment]
# TODO: find better solution than increasing max_flowcells all the time. I can't find anywhere what we do with flowcells with status `removed`, @patrikgrenfeldt am I missing something? We should define and implement some form of clean up functionality.

**Comment by barrystokman** [comment]
@moonso @patrikgrenfeldt I've updated this old and approved PR to solve an issue in prod and it would help @moahaegglund if we could deploy this and implement a more durable fix later.

**Comment by moonso** [comment]
Sure, I have stayed away from reviewing WIP PRs until the WIP-flag is removed. But I was not sure how you are working with that flag.

But this is ready to be reviewed now?

File: cg/cli/backup.py [hunk 3]
```diff
@@ -18,25 +18,30 @@ def backup(context: click.Context):
 
 @backup.command("fetch-flowcell")
 @click.option("-f", "--flowcell", help="Retreive a specific flowcell")
-@click.option("--dry", is_flag=True, help="Only log, do not contact PDC")
+@click.option(
+    "--dry-run", "dry_run", is_flag=True, help="Don't retrieve from PDC or set flowcell's status"
```
Comment (moonso) [inline cg/cli/backup.py:22]:
```suggestion
    "--dry-run", is_flag=True, help="Don't retrieve from PDC or set flowcell's status"
```
This is not needed here so I removed it

File: cg/cli/backup.py [hunk 4]
```diff
@@ -18,25 +18,30 @@ def backup(context: click.Context):
...
+)
 @click.pass_context
-def fetch_flowcell(context: click.Context, dry: bool, flowcell: str):
+def fetch_flowcell(context: click.Context, dry_run: bool, flowcell: str):
     """Fetch the first flowcell in the requested queue from backup."""
     status_api = Store(context.obj["database"])
+    max_flowcells = context.obj.get("max_flowcells", 1250)
```
Comment (moonso) [inline cg/cli/backup.py:28]: Put the default in a global variable? Perhaps with a comment about the choice of 1250
```suggestion
    max_flowcells = context.obj.get("max_flowcells", MAX_FLOWCELLS)
```

File: cg/cli/backup.py [hunk 5]
```diff
@@ -18,25 +18,30 @@ def backup(context: click.Context):
...
         if flowcell_obj is None:
             LOG.error(f"{flowcell}: not found in database")
             context.abort()
     else:
         flowcell_obj = None
-    retrieval_time = backup_api.fetch_flowcell(flowcell_obj=flowcell_obj, dry=dry)
+
+    retrieval_time = backup_api.fetch_flowcell(flowcell_obj=flowcell_obj, dry_run=dry_run)
+
     if retrieval_time is None:
```
Comment (moonso) [inline cg/cli/backup.py:41]: I would suggest to flatten this section by introducing some
```python
if <condition>:
    continue
```

Comment (moonso) [reply cg/cli/backup.py:41]: This section would look like
```python
if retrieval_time:
    hours = retrieval_time / 60 / 60
    LOG.info(f"Retrieval time: {hours:.1}h")
    return

if not flowcell:
    return

LOG.info(f"{flowcell}: updating flowcell status to requested")
if not dry_run:
    flowcell_obj.status = "requested"
    status_api.commit()
```

File: cg/meta/backup.py [hunk 6]
```diff
@@ -12,17 +12,18 @@
...
-    def maximum_flowcells_ondisk(self, max_flowcells: int = 1000) -> bool:
+    def maximum_flowcells_ondisk(self) -> bool:
         """Check if there's too many flowcells already "ondisk"."""
         ondisk_flowcells = self.status.flowcells(status="ondisk").count()
         LOG.debug("ondisk flowcells: %s", ondisk_flowcells)
-        return ondisk_flowcells > max_flowcells
+        return ondisk_flowcells > self.max_flowcells
 
-    def check_processing(self, max_flowcells: int = 3) -> bool:
+    def check_processing(self, max_flowcells: int = 1) -> bool:
```
Comment (moonso) [inline cg/meta/backup.py:26]: This is confusing, are we using the same variable name with a different meaning here? Or why not use `self.max_flowcells`?

File: cg/meta/backup.py [hunk 7]
```diff
@@ -53,21 +54,22 @@ def fetch_flowcell(self, flowcell_obj: models.Flowcell = None, dry: bool = False
...
 
-        if flowcell_obj is None:
+        if not dry_run and flowcell_obj is None:
             flowcell_obj = self.pop_flowcell()
             if flowcell_obj is None:
                 LOG.info("no flowcells requested")
                 return None
 
-        LOG.info("%s: retreiving from PDC", flowcell_obj.name)
+        if not dry_run:
```
Comment (moonso) [inline cg/meta/backup.py:63]: Why would we not log this even if it is a dryrun?

File: cg/meta/backup.py [hunk 8]
```diff
@@ -39,7 +40,7 @@ def pop_flowcell(self) -> models.Flowcell:
             self.status.commit()
         return flowcell_obj
 
-    def fetch_flowcell(self, flowcell_obj: models.Flowcell = None, dry: bool = False):
```
Comment (moonso) [inline cg/meta/backup.py:42]: Returns what?

File: cg/meta/backup.py (hunk 6 above)
Comment (barrystokman) [reply cg/meta/backup.py:26]: I'm not sure, this is not my code. I only changed the number from 1000 to 1250 :)

You have a good point of course and I'll have a look.

**Comment by barrystokman** [comment]
@moonso I'll address all your comments. I haven't even looked at the code besides the number that I changed since patrik had already approved, and as I said it's not my code to begin with.

**Comment by moonso** [comment]
Ok sorry about that. But I think they are relevant!

At least it was a small PR :)

**Comment by barrystokman** [comment]
> But I think they are relevant!

I agree! :)

File: cg/cli/backup.py (hunk 5 above)
Comment (barrystokman) [reply cg/cli/backup.py]: fixed in https://github.com/Clinical-Genomics/cg/pull/397/commits/e1bb9c2bed2c1873693e9e31ff13f794a944ead3

File: cg/meta/backup.py (hunk 6 above)
Comment (barrystokman) [reply cg/meta/backup.py]: Answer: we're dealing with two different max_flowcells! one for the max number on disk, and on for the max number being processed. https://github.com/Clinical-Genomics/cg/pull/397/commits/eda51b768f6bc26eac7149720c70ba096a80d29e will make it more clear which is which :)

File: cg/meta/backup.py (hunk 7 above)
Comment (barrystokman) [reply cg/meta/backup.py:63]: Fixed here https://github.com/Clinical-Genomics/cg/pull/397/commits/e272b84fed676d2d759b0f83e2dac589050b8d04

File: cg/meta/backup.py (hunk 8 above)
Comment (barrystokman) [reply cg/meta/backup.py:42]: type hint added here: https://github.com/Clinical-Genomics/cg/pull/397/commits/6ab1a7e57f0764aa7613994b0b9231f716d360f3

**Comment by barrystokman** [comment]
Retest:

TC1: flowcells on disk < max (1250) but > old max (1000) and one flowcell is already `processing`
Expected outcome: cli command exits with message 'processing queue is full'

Number of flowcells ondisk is greater than the old MAX_FLOWCELLS_ON_DISK of 1000:
image
:ballot_box_with_check: : 
One or more flowcells have status `processing`:
image
:ballot_box_with_check: 

Run `cg backup fetch-flowcell`:
image
:heavy_check_mark: 

TC2: flowcells on disk < max (1250) but > old max (1000) and processing queue is empty:
image
:ballot_box_with_check: : 
No flowcells have status `processing`:
image
:ballot_box_with_check: : 

Run `cg backup fetch-flowcell`:
image

The flowcell is now in the processing queue to be retrieved from PDC:
image
:heavy_check_mark: 

TC3: as TC1, but ran with `--dry-run` option:
image
:heavy_check_mark:

TC4: as TC2, but ran with `--dry-run` option:
image
:negative_squared_cross_mark:

**Comment by barrystokman** [comment]
@moonso I've addresses your comments and retested the whole thing, do you think it'd ready to be deployed?

File: cg/cli/backup.py [hunk 9]
```diff
@@ -7,6 +7,7 @@
 from cg.meta.backup import BackupApi
 
 LOG = logging.getLogger(__name__)
+MAX_FLOWCELLS_ON_DISK = 1250  # Increased by 250 when the previous limit of 1000 was reached
```
Comment (moonso) [inline cg/cli/backup.py:10]: 👍

File: cg/meta/backup.py [hunk 10]
```diff
@@ -12,34 +12,36 @@
...
-    def maximum_flowcells_ondisk(self, max_flowcells: int = 1000) -> bool:
+    def maximum_flowcells_ondisk(self) -> bool:
         """Check if there's too many flowcells already "ondisk"."""
         ondisk_flowcells = self.status.flowcells(status="ondisk").count()
         LOG.debug("ondisk flowcells: %s", ondisk_flowcells)
-        return ondisk_flowcells > max_flowcells
+        return ondisk_flowcells > self.max_flowcells_on_disk
 
-    def check_processing(self, max_flowcells: int = 3) -> bool:
+    def check_processing(self, max_processing_flowcells: int = 1) -> bool:
```
Comment (moonso) [inline cg/meta/backup.py:26]: Much better :)

**Comment by barrystokman** [comment]
Bumped:
image


Deployed:
image

**Pull request merged**